src/lib.rs

```diff
@@ -102,20 +104,148 @@ pub struct SelectionOutput {
 pub fn select_coin_bnb(
     inputs: &[OutputGroup],
     options: CoinSelectionOpt,
+    rng: &mut ThreadRng,
 ) -> Result<SelectionOutput, SelectionError> {
-    unimplemented!()
+    let mut selected_inputs: Vec<usize> = vec![];
+
+    const BNB_TRIES: u32 = 1_000_000;
+
+    let mut sorted_inputs: Vec<(usize, OutputGroup)> = inputs
+        .iter()
+        .enumerate()
+        .map(|(index, input)| (index, *input))
+        .collect();
+    sorted_inputs.sort_by_key(|(_, input)| std::cmp::Reverse(input.value));
+
+    let bnb_selected_coin = bnb(
+        &sorted_inputs,
+        &mut selected_inputs,
+        0,
+        0,
+        BNB_TRIES,
+        &options,
+        rng,
+    );
+    match bnb_selected_coin {
+        Some(selected_coin) => {
+            let accumulated_value: u64 = selected_coin
+                .iter()
+                .fold(0, |acc, &i| acc + inputs[i].value);
+            let accumulated_weight: u32 = selected_coin
+                .iter()
+                .fold(0, |acc, &i| acc + inputs[i].weight);
+            let estimated_fee = 0;
+            let waste = calculate_waste(
+                inputs,
+                &selected_inputs,
+                &options,
+                accumulated_value,
+                accumulated_weight,
+                estimated_fee,
+            );
+            let selection_output = SelectionOutput {
+                selected_inputs: selected_coin,
+                waste: WasteMetric(waste),
+            };
+            Ok(selection_output)
+        }
+        None => Err(SelectionError::NoSolutionFound),
+    }
 }
 
 /// Return empty vec if no solutions are found
+// changing the selected_inputs : &[usize] -> &mut Vec<usize>
 fn bnb(
     inputs_in_desc_value: &[(usize, OutputGroup)],
-    selected_inputs: &[usize],
-    effective_value: u64,
+    selected_inputs: &mut Vec<usize>,
+    acc_eff_value: u64,
     depth: usize,
     bnp_tries: u32,
     options: &CoinSelectionOpt,
-) -> Vec<usize> {
-    unimplemented!()
+    rng: &mut ThreadRng,
+) -> Option<Vec<usize>> {
+    let target_for_match = options.target_value
```

Comment:
This value could be computed only once as it depends on the options
No need to recompute this during every recursive call.
This can be included in the function parameter